### PR TITLE
Support getting used memory from a user perspective at `PooledByteBufAllocator` level

### LINKED ISSUE #11637: Add support getting the pooled memory size for each PooledByteBufAllocator instance
### Motivation
Our server has its own implementation to decide whether to provide service for users according to the real-time total free memory of the heap memory and the direct memory (mainly the memory used by `io.netty.buffer.PooledByteBufAllocator` for our server). If our server detects it's running out of physical memory, it will just respond with the client requests with something like `Service Unavailable`.

But the problem is: We cannot know the free memory of PooledByteBufAllocator efficiently. For example, we have allocated and cached 8GB chunks in a `PooledByteBufAllocator`, but there is no property to know: what exactly the free memory is, 1GB or 8GB?

### Solution
The [PR](https://github.com/netty/netty/pull/5133/files) in 2016 is similar to what we want and it provides a new property `pooledBytes` to record the free memory of the arena but we want the free memory in the scope of each allocator instance.

And `io.netty.maxDirectMemory` is similar but isn't what we want because it applies the limit to all buffers while we just want to know the free memory size of a PooledByteBufAllocator instance

### PR DESCRIPTION
**Pull request by JamesChenX**
## Motivation
Our server has its own implementation to decide whether to provide service for users according to the real-time estimated total free memory of the heap memory and the direct memory (mainly the memory used by `io.netty.buffer.PooledByteBufAllocator` for our server). If our server detects it's running out of physical memory, it will just respond with the client requests with something like `Service Unavailable`.

But the problem is: We cannot know the used memory of PooledByteBufAllocator efficiently. For example, we have allocated and cached 8GB chunks in a `PooledByteBufAllocator`, but there is no property to know: From the user perspective, what exactly the used memory is, 1GB or 8GB?

The PR will provide `usedDirectBytes` and `usedHeapBytes` for these purposes.

## Modification
(Honestly, I didn't come up with better names to distinguish the existing `usedDirectMemory` and the new `usedDirectBytes`. If Netty were a new project, we can distinguish these memories by calling the existing `usedDirectMemory` `committedDirectMemory` and the new memory metric `usedDirectMemory`, but Netty has been widely used in the Java world so I don't call them in this way)

When a pooled buffer is allocated/deallocated, grows, and shrinks, count the increase/decrease size to count `usedDirectBytes` and `usedHeapBytes`

## Result

Fixes #11637

### CODE REVIEW

**Comment by chrisvest** [comment]
Thanks, @JamesChenX. There are some check style issues. Also, please make sure you've signed the ICLA: https://netty.io/wiki/developer-guide.html

**Comment by JamesChenX** [comment]
@chrisvest Thanks for the review, and I have signed the ICLA. PTAL.

**Comment by JamesChenX** [comment]
@medusar `usedHeapMemory` and `usedDirectMemory` count the cached memory in `PooledByteBufAllocator` while what we want is to count the used memory in the pooled memory from the user perspective. I know it's a little confusing and that's why I am hesitating about what names we should use to distinguish these two kinds of memory.

e.g.
```java
ByteBuf buffer = allocator.directBuffer(16, 128);
assertEquals(16, allocator.usedDirectBytes());
assertEquals(16777216, allocator.usedHeapMemory());

buffer.release();
assertEquals(0, allocator.usedDirectBytes());
assertEquals(16777216, allocator.usedHeapMemory());
```

File: buffer/src/main/java/io/netty/buffer/PooledByteBuf.java [hunk 1]
```diff
@@ -98,6 +98,11 @@ public final ByteBuf capacity(int newCapacity) {
             return this;
         }
         checkNewCapacity(newCapacity);
+        if (isDirect()) {
+            chunk.arena.parent.incrementUsedDirectBytes(newCapacity - length);
+        } else {
+            chunk.arena.parent.incrementUsedHeapBytes(newCapacity - length);
+        }
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/PooledByteBuf.java:105]: This logic should move into `PoolArena.reallocate`.
At this location, if the new capacity is less than the current capacity, we would incorrectly count the used memory down even though the memory wasn't actually returned to the pool.

File: buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java [hunk 2]
```diff
@@ -382,6 +385,8 @@ protected ByteBuf newHeapBuffer(int initialCapacity, int maxCapacity) {
                     new UnpooledHeapByteBuf(this, initialCapacity, maxCapacity);
         }
 
+        incrementUsedHeapBytes(initialCapacity);
```
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:388]: I think we should use the size of the actual memory consumed by the buffer, rather than the sum of the capacities of all in-use buffers. So these increments and decrements will be in terms of pool chunk sub-page element size (for "small" allocation), pool chunk page-multiple run size (for "normal" allocations), or the requested capacity (for "huge" allocations).

Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java:388]: In other words, I'd like to see this accounting be pushed down to where memory is removed from, or returned to, the pool.

**Review by chrisvest — CHANGES_REQUESTED** [review]
Hey, thanks for doing this work.

I'm not in favour of the granularity of the accounting being done here, because it misrepresents the amount of memory that is actually removed from the shared pool by a buffer. I think it should be done in terms of the internal allocation units.

**Comment by JamesChenX** [comment]
@chrisvest Thanks. I think we need to confirm whether our expectations are the same for the new methods `getUsedDirectBytes` and `getUsedHeapBytes` first.
What you have described seems the behavior of the existing `getUsedDirectMemory` and `getUsedHeapMemory`, which count the real used and cached buffers in PooledByteBufAllocator. But the new methods `usedDirectBytes` and `usedHeapBytes` just count the used bytes that follow the lifecycle of a pooled buffer, which is what we need in our project. e.g.:
```java
ByteBuf buffer = allocator.directBuffer(16, 128);
assertEquals(16, allocator.usedDirectBytes()); // ok
assertEquals(16777216, allocator.usedDirectMemory()); // ok. "usedDirectMemory" should named "committedDirectMemory"

buffer.release();
assertEquals(0, allocator.usedDirectBytes()); // ok
assertEquals(16777216, allocator.usedDirectMemory()); // ok
```
My expectation for these two new methods is the asserts in: https://github.com/netty/netty/pull/11642/files#diff-5356c1287108c4b93aee8b54109d5ef4d4b9cb2179af1806384e0c520aa9dfe6R694-R722

Do we expect the same thing?

**Comment by chrisvest** [comment]
@JamesChenX We can think of it as memory committed by the pool, and memory committed to a buffer instance, if that helps.

The existing `allocator.usedDirectMemory()` (for example) computes a sum of `arena.numActiveBytes()`, and that in turn computes a sum of `m.chunkSize()` where `m` is a `PoolChunk`. That's the memory committed to the pool, because each chunk is (for direct memory) a single direct memory allocation.

Chunks are then used in pieces by many buffers. Each direct buffer gets a section of a chunk committed to it, which is greater-than-or-equal-to the buffer size, so the buffer can be smaller than the memory committed to it. I think the new method should return the sum of memory committed to the buffers.

I think we can get that with code similar to what's done for `allocator.usedDirect/HeapMemory()`, but instead of using sums of `chunk.chunkSize()`, we use `chunk.chunkSize() - chunk.freeBytes()`.

So in summary, I think we had slightly different expectations, but if the purpose is to do respond to high utilisation of the pooled memory resources, then going by the memory committed to buffers rather than their sizes, will be more accurate.

Makes sense?

Then, once we are getting the values out that we want, we can see if computing these sums is too slow and if it would be better to have some `LongCounters`, or perhaps per-arena `AtomicLongs`, to speed up querying.

**Comment by JamesChenX** [comment]
@chrisvest Thanks for the explanation and I see your point. Though we expect different metrics, `chunk.chunkSize() - chunk.freeBytes()` is accurate and performant enough for our scenarios. So I close the PR, Thanks!
